File: HeapSort.pl
#!/usr/bin/perl
use strict;
use warnings;
use Data::Dumper;

print "Input Number of Data:";
my $num = <STDIN>;

my @array = map { int rand $num } 0..$num;
our $heap_size = @array;

&heapsort(\@array);

exit;

sub parent {

	my ($i) = @_;
	return int(($i-1)/2);

}

sub left {

	my ($i) = @_;
	return 2*$i+1;

}

sub right {

	my ($i) = @_;
	return 2*($i+1);
}

sub max_heapify {

	my ($array_ref,$i) = @_;
	
	my $l = &left($i);
	my $r = &right($i);
	my $largest = $i;

	print "Searching $i\n";

	if($l < $heap_size && $array_ref->[$l] > $array_ref->[$i]){
		
		print "Hit L\n";
		$largest = $l;

	}

	if($r < $heap_size && $array_ref->[$r] > $array_ref->[$l]){

		print "Hit R\n";
		$largest = $r;

	}

	if($largest != $i){
		print "Swap $largest $i Value $array_ref->[$largest] $array_ref->[$i]\n";
		($array_ref->[$largest],$array_ref->[$i]) = ($array_ref->[$i],$array_ref->[$largest]);
		&max_heapify($array_ref,$largest);
	}

}

sub build_max_heap {

	my ($array_ref) = @_;

	for(my $i = int $heap_size/2; int $i > 0; $i--){

		&max_heapify($array_ref,$i-1);
	
	}

}

sub heapsort {

	my ($array_ref) = @_;

	print "Input Data\n";
	print Dumper $array_ref;

	&build_max_heap($array_ref); 

	print "Max Heapified\n";
	print Dumper $array_ref;

	for(my $i = $#$array_ref; $i > 0; $i--){

		print "Processing...\n";
		print Dumper $array_ref;
		($array_ref->[0],$array_ref->[$i]) = ($array_ref->[$i],$array_ref->[0]);
		$heap_size--;
		&max_heapify($array_ref,0);
		
	}

	print "Sorted\n";
	print Dumper $array_ref;

}

File: HeapSort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int heap_size;
int num;

void swap(int *a,int *b){

	int temp;
	temp = *a;
	*a = *b;
	*b = temp;
	return;
}
int left(int i) {

	return 2*i+1;

}

int right(int i) {

	return 2*(i+1);
}

int max_heapify(int* p,int i) {

	int l = left(i);
	int r = right(i);
	int largest = i;

	if(l < heap_size && *(p+l) > *(p+i)){
		
		largest = l;

	}

	if(r < heap_size && *(p+r) > *(p+l)){

		largest = r;

	}

	if(largest != i){
		swap(p+largest,p+i);
		max_heapify(p,largest);
	}

	return 0;

}

int build_max_heap(int* p) {

	int i;
	for(i = heap_size/2; i > -1; i--){

		max_heapify(p,i);
	
	}

	return 0;

}

int heapsort(int *p) {

	int i;
	build_max_heap(p); 

	for(i = num-1; i > -1; i--){

		swap(p,p+i);
		heap_size--;
		max_heapify(p,0);
		
	}

	return 0;

}

int main(){

	printf("Input Number of Data:");

	scanf("%d",&num);

	int* p;
	p  = malloc(sizeof(int)*num);

	srand((unsigned)time(NULL));

	int i;
	
	printf("Input_Array\n");
	for(i = 0; i < num; i++){

		*(p+i) =  rand()%num;
		printf("%d\n",*(p+i));
	
	}
	heap_size = num;

	
	heapsort(p);

	printf("Sorted_Array\n");
	for(i = 0; i < num; i++){
		
		printf("%d\n",*(p+i));
	
	}

	
	return 0;
}


